src/lib/storage/vector_compression/fixed_size_bit_aligned/fixed_size_bit_aligned_compressor.cpp

```diff
@@ -0,0 +1,42 @@
+#include "fixed_size_bit_aligned_compressor.hpp"
+
+#include <algorithm>
+#include <cmath>
+
+#include "compact_vector.hpp"
+
+namespace opossum {
+
+class FixedSizeBitAlignedVector;
+
+std::unique_ptr<const BaseCompressedVector> FixedSizeBitAlignedCompressor::compress(
+    const pmr_vector<uint32_t>& vector, const PolymorphicAllocator<size_t>& alloc,
+    const UncompressedVectorInfo& meta_info) {
+  const auto max_value = _find_max_value(vector);
+  const auto required_bits = _get_required_bits(max_value);
+
+  auto data = pmr_compact_vector<uint32_t>(required_bits, alloc);
+  // resize to avoid allocating too much memory with auto-growing vector. Doesn't support reserve().
+  data.resize(vector.size());
+  std::copy(vector.cbegin(), vector.cend(), data.begin());
+
+  return std::make_unique<FixedSizeBitAlignedVector>(std::move(data));
+}
+
+std::unique_ptr<BaseVectorCompressor> FixedSizeBitAlignedCompressor::create_new() const {
+  return std::make_unique<FixedSizeBitAlignedCompressor>();
+}
+
+uint32_t FixedSizeBitAlignedCompressor::_find_max_value(const pmr_vector<uint32_t>& vector) const {
+  const auto it = std::max_element(vector.cbegin(), vector.cend());
+  return it != vector.cend() ? *it : 0u;
+}
+
+uint32_t FixedSizeBitAlignedCompressor::_get_required_bits(uint32_t max_value) const {
+  if (max_value == 0u) {
+    return 1u;
+  }
+  return static_cast<uint32_t>(std::ceil(log2(max_value + 1u)));
```

Comment:
Why +1?

Reply:
we add 1 to max_value because log2(1) = 0 but we need one bit to represent it.
added comment explaining this.

Reply:
But that might hurt quite often for columns with very few values, wouldn’t it?
What about `max(1,...)`?

Reply:
those are the return values for max value 1-16:
```
0: 1
1: 1
2: 2
3: 2
4: 3
5: 3
6: 3
7: 3
8: 4
9: 4
10: 4
11: 4
12: 4
13: 4
14: 4
15: 4
16: 5
```
What do you mean, where does it "hurt"?

Reply:
but yes, you could argue to write a test for that

Reply:
Sorry, I mixed something up. Nvm.

All good with a simple short commment